subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -165,15 +165,30 @@ public void visitOutgoingDependencies(Collection<EdgeState> target) {
             removeOutgoingEdges();
         }
 
-        for (DependencyMetadata dependency : metaData.getDependencies()) {
-            if (isExcluded(resolutionFilter, dependency)) {
-                continue;
+        visitDependencies(resolutionFilter, optionalDependenciesHandler, target);
+
+    }
+
+    protected void visitDependencies(ModuleExclusion resolutionFilter, OptionalDependenciesHandler optionalDependenciesHandler, Collection<EdgeState> resultingOutgoingEdges) {
+        boolean isOptionalConfiguration = "optional".equals(metaData.getName());
```

Comment:
⭕️ I wonder why we have this stateful optional dependencies handler instead of creating it here. It seems more like a data structure, not so much like a service. The benefit would be that you can have two different implementations, one that's a no-op and one that actually does something. Then you don't need a boolean flag (isOptionalConfiguration). And you wouldn't need start/stop or try/finally either, as the state would be local to this method.

Reply:
This is a performance optimization. We visit outgoing edges very often, so it's about avoiding creating an object and dropping it many times.